h5py/_selector.pyx

```diff
@@ -200,6 +228,58 @@ cdef class Reader:
             efree(tmp_start)
             efree(tmp_count)
 
+
+    def make_selection(self, tuple args):
+        """Apply indexing/slicing args and create a high-level selection object
+
+        Returns an instance of SimpleSelection or FancySelection, with a copy
+        of the selector's dataspace.
+        """
+        cdef:
+            SpaceID space
+            tuple shape, start, count, step, scalar, arr_shape
+            int arr_rank, i
+            npy_intp* arr_shape_p
+
+        self.apply_args(args)
+        space = SpaceID(H5Scopy(self.space))
+
+        shape = convert_dims(self.dims, self.rank)
+        count = convert_dims(self.count, self.rank)
+
+        from ._hl.selections import SimpleSelection, FancySelection
+
+        if self.is_fancy:
+            arr_shape = tuple(
+                int(self.count[i]) for i in range(self.rank) if self.scalar[i]
+            )
+            return FancySelection(shape, space, count, arr_shape)
+        else:
+            start = convert_dims(self.start, self.rank)
+            step = convert_dims(self.stride, self.rank)
+            scalar = convert_bools(self.scalar, self.rank)
+            return SimpleSelection(shape, space, (start, count, step, scalar))
+
+
+cdef class Reader:
+    cdef hid_t dataset
+    cdef Selector selector
+    cdef TypeID h5_memory_datatype
+    cdef int np_typenum
+
+    def __cinit__(self, DatasetID dsid):
+        self.dataset = dsid.id
+        self.selector = Selector(dsid.get_space())
```

Comment:
Maybe rename this `self._selector`?  This is definitely part of the private API (given that we mutate it `read`).

I am normally much more composition over inheritance, but does sub-classing `Reader` from `Selector` make sense here given that they used to be one class?  I am 75% on the side of leaving the change as-is, but thinking out-loud.

Reply:
It's not accessible from Python, because it's a cdef-ed attribute not marked as public. I've not been worrying about third party Cython code trying to use it - do you think we should?

I wondered about inheriting as well. I leant towards composition, because a `Reader` knows about an extra dimension of things (datasets and datatypes), whereas a `Selector` operates simply on a dataspace. Also, `__cinit__` doesn't quite behave like a normal Python method:

> If your extension type has a base type, any existing __cinit__() methods in the base type hierarchy are automatically called before your __cinit__() method.

([Cython docs])

So it wouldn't be easy for `Reader` defined as a subclass to accept a DatasetID instead of a SpaceID. There are bound to be ways around that, but it makes things less neat.

Reply:
I was thinking of this as public to python, but given that it is not easily accessible, I have no concerns.  Could @kif weigh in on if we need to worry about the cython side extension?

I on fully on board with leaving it as composition.